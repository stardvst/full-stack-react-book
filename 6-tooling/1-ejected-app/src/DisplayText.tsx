import React, { FC, useState } from 'react';

interface DisplayTextProps {
  getUserFullName: (username: string) => Promise<string>;
}

const DisplayText: FC<DisplayTextProps> = ({ getUserFullName }) => {
  const [text, setText] = useState('');
  const [message, setMessage] = useState('');
  const [todos, setTodos] = useState<Array<JSX.Element>>();

  const setUserTodos = async () => {
    const userResponse = await fetch('https://jsonplaceholder.typicode.com/users');
    if (userResponse.ok) {
      const users = await userResponse.json();
      const userByName = users.find((user: any) => user.username.toLowerCase() === text);
      console.log('user by username:', userByName);
      //return userByName.name;

      const todosResponse = await fetch('https://jsonplaceholder.typicode.com/todos');
      if (todosResponse.ok) {
        const todos = await todosResponse.json();
        const userTodos = todos.filter((todo: any) => todo.userId === userByName.id);
        console.log('user todos:', userTodos);

        const todoList = userTodos.map((todo: any) => <li key={todo.id}>{todo.title}</li>);
        setTodos(todoList);
      }
    }
  };

  const onChangeText = (e: React.ChangeEvent<HTMLInputElement>) => {
    setText(e.target.value);
  };

  const onClickShowMessage = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    setMessage(text.length ? `welcome to react testing, ${await getUserFullName(text)}` : '');
    setUserTodos();
  };

  return (
    <form>
      <div>
        <label htmlFor="name">enter your name:</label>
        <input type="text" value={text} onChange={onChangeText} data-testid="name-input" />
        <button data-testid="name-submit" onClick={onClickShowMessage}>
          show message
        </button>
      </div>
      <div>
        <p data-testid="final-msg">{message}</p>
      </div>
      <div>this is just a test entry</div>
      <ul style={{ marginTop: '1rem', listStyleType: 'none' }}>{todos}</ul>
    </form>
  );
};

export default DisplayText;
