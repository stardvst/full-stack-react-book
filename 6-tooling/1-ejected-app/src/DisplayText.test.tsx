import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import DisplayText from './DisplayText';

describe('test displaytext', () => {
  it('renders without crashing', () => {
    const { baseElement } = render(<DisplayText />);
    console.log(baseElement.innerHTML);
    expect(baseElement).toBeInTheDocument();
  });

  it('receives input text', () => {
    const text = 'testuser';
    const { getByTestId } = render(<DisplayText />);

    const input = getByTestId('name-input');
    fireEvent.change(input, { target: { value: text } });

    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(text);
  });

  it('show welcome message', () => {
    const text = 'testuser';
    const message = `welcome to react testing, ${text}`;

    const { getByTestId } = render(<DisplayText />);
    const input = getByTestId('name-input');
    const label = getByTestId('final-msg');

    fireEvent.change(input, { target: { value: text } });
    const btn = getByTestId('name-submit');
    fireEvent.click(btn);

    expect(label).toBeInTheDocument();
    expect(label.innerHTML).toBe(message);
  });

  it('matches snapshot', () => {
    const { baseElement } = render(<DisplayText />);
    expect(baseElement).toMatchSnapshot();
  });
});
