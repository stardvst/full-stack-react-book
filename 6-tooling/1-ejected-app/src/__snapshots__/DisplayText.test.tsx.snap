// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test displaytext matches snapshot 1`] = `
<body>
  <div>
    <form>
      <div>
        <label
          for="name"
        >
          enter your name:
        </label>
        <input
          data-testid="name-input"
          type="text"
          value=""
        />
        <button
          data-testid="name-submit"
        >
          show message
        </button>
      </div>
      <div>
        <p
          data-testid="final-msg"
        />
      </div>
      <div>
        this is just a test entry
      </div>
    </form>
  </div>
</body>
`;
